File: Operating_Systems/CPU-Scheduling/RoundRobin.py
# implement a round robin scheduling algorithm

from collections import deque

class Process:
    def __init__(self, name, arrival, burst):
        self.name = name
        self.arrival = arrival
        self.burst = burst
        self.remaining_time = burst
        self.completion = 0
        self.tat = 0
        self.wt = 0

def round_robin(processes, quantum):
    # Sort by Arrival Time
    processes.sort(key=lambda x: x.arrival)
    
    # Initialize variables
    time = 0
    n = len(processes)
    ready_queue = deque()
    results = []
    
    while True:
        # Add processes that have arrived to the ready queue
        for p in processes:
            if p.arrival <= time and p.remaining_time > 0:
                ready_queue.append(p)
        
        # If ready queue is empty, break
        if not ready_queue:
            break
        
        # Process the first process in the ready queue
        current_process = ready_queue.popleft()
        if current_process.remaining_time > quantum:
            time += quantum
            current_process.remaining_time -= quantum
            ready_queue.append(current_process)
        else:
            time += current_process.remaining_time
            current_process.remaining_time = 0
            current_process.completion = time
            current_process.tat = current_process.completion - current_process.arrival
            current_process.wt = current_process.tat - current_process.burst
            results.append(current_process)
    
    return results

# Example Processes
processes = [
    Process('P1', 0, 10),
    Process('P2', 1, 4),
    Process('P3', 2, 5),
    Process('P4', 3, 3)
]

quantum = 2

results = round_robin(processes, quantum)

print("Process | Arrival | Burst | Completion | Turnaround | Waiting")
for p in results:
    print(f"{p.name:>7} | {p.arrival:>7} | {p.burst:>5} | {p.completion:>10} | {p.tat:>10} | {p.wt:>7}")

    

File: Operating_Systems/CPU-Scheduling/RoundRobin.js
// implement a round robin scheduling algorithm
// The round-robin scheduling algorithm is designed especially for time-sharing systems. It is similar to the first-come, first-served scheduling algorithm, but preemption is added to switch between processes. A small unit of time, called a time quantum or time slice, is defined. A time quantum is generally from 10 to 100 milliseconds. The ready queue is treated as a circular queue. The CPU scheduler goes around the ready queue, allocating the CPU to each process for a time interval of up to one time quantum. To implement the round-robin scheduling algorithm, we will use a queue data structure.

class Queue {
    constructor() {
        this.items = [];
    }
    
    enqueue(element) {
        this.items.push(element);
    }
    
    dequeue() {
        if (this.isEmpty()) return "Underflow";
        return this.items.shift();
    }
    
    front() {
        if (this.isEmpty()) return "No elements in Queue";
        return this.items[0];
    }
    
    isEmpty() {
        return this.items.length === 0;
    }
    
    printQueue() {
        let str = "";
        for (let i = 0; i < this.items.length; i++) str += this.items[i] + " ";
        return str;
    }
    }

    class RoundRobin {
    constructor() {
        this.queue = new Queue();
    }

    findWaitingTime(processes, n, bt, wt, quantum) {
        let rem_bt = [];
        for (let i = 0; i < n; i++) rem_bt[i] = bt[i];
        let t = 0;
        while (true) {
            let done = true;
            for (let i = 0; i < n; i++) {
                if (rem_bt[i] > 0) {
                    done = false;
                    if (rem_bt[i] > quantum) {
                        t += quantum;
                        rem_bt[i] -= quantum;
                    } else {
                        t = t + rem_bt[i];
                        wt[i] = t - bt[i];
                        rem_bt[i] = 0;
                    }
                }
            }
            if (done === true) break;
        }
    }

    findTurnAroundTime(processes, n, bt, wt, tat) {
        for (let i = 0; i < n; i++) tat[i] = bt[i] + wt[i];
    }

    findavgTime(processes, n, bt, quantum) {
        let wt = [],
            tat = [];
        this.findWaitingTime(processes, n, bt, wt, quantum);
        this.findTurnAroundTime(processes, n, bt, wt, tat);
        let total_wt = 0,
            total_tat = 0;
        for (let i = 0; i < n; i++) {
            total_wt += wt[i];
            total_tat += tat[i];
        }
        return {
            avg_wt: total_wt / n,
            avg_tat: total_tat / n,
        };
    }

    }

    // Implementation
    const processes = [1, 2, 3];
    const n = processes.length;
    const burst_time = [10, 5, 8];
    const quantum = 2;
    const roundRobin = new RoundRobin();
    const result = roundRobin.findavgTime(processes, n, burst_time, quantum);
    console.log(`Average Waiting Time: ${result.avg_wt}`);
    console.log(`Average Turnaround Time: ${result.avg_tat}`);
